<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="ccbr1271_ERVpipeline Pipeline"><meta name=author content="Vishal Koparde, Ph.D."><link href=../usage/ rel=prev><link rel=icon href=../assets/images/favicon.png><meta name=generator content="mkdocs-1.5.3, mkdocs-material-9.5.6"><title>Resources - ccbr1271_ERVpipeline Pipeline</title><link rel=stylesheet href=../assets/stylesheets/main.50c56a3b.min.css><link rel=stylesheet href=../assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#genomes class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=.. title="ccbr1271_ERVpipeline Pipeline" class="md-header__button md-logo" aria-label="ccbr1271_ERVpipeline Pipeline" data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> ccbr1271_ERVpipeline Pipeline </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Resources </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=red data-md-color-accent=red aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg> </label> </form> <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/CCBR/ccbr1271_ERVpipeline title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> CCBR/ccbr1271_ERVpipeline </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=.. class=md-tabs__link> Background </a> </li> <li class=md-tabs__item> <a href=../flowchart/ class=md-tabs__link> Flowchart </a> </li> <li class=md-tabs__item> <a href=../usage/ class=md-tabs__link> Usage </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=./ class=md-tabs__link> Resources </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=.. title="ccbr1271_ERVpipeline Pipeline" class="md-nav__button md-logo" aria-label="ccbr1271_ERVpipeline Pipeline" data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg> </a> ccbr1271_ERVpipeline Pipeline </label> <div class=md-nav__source> <a href=https://github.com/CCBR/ccbr1271_ERVpipeline title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> CCBR/ccbr1271_ERVpipeline </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=.. class=md-nav__link> <span class=md-ellipsis> Background </span> </a> </li> <li class=md-nav__item> <a href=../flowchart/ class=md-nav__link> <span class=md-ellipsis> Flowchart </span> </a> </li> <li class=md-nav__item> <a href=../usage/ class=md-nav__link> <span class=md-ellipsis> Usage </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Resources </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Resources </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#genomes class=md-nav__link> <span class=md-ellipsis> Genomes </span> </a> </li> <li class=md-nav__item> <a href=#star-index class=md-nav__link> <span class=md-ellipsis> STAR index </span> </a> </li> <li class=md-nav__item> <a href=#geve-annotations class=md-nav__link> <span class=md-ellipsis> GEVE annotations </span> </a> </li> <li class=md-nav__item> <a href=#creating-taxid-to-lineage-lookup class=md-nav__link> <span class=md-ellipsis> Creating taxid-to-lineage lookup </span> </a> <nav class=md-nav aria-label="Creating taxid-to-lineage lookup"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#download-dmp-files class=md-nav__link> <span class=md-ellipsis> Download .dmp files </span> </a> </li> <li class=md-nav__item> <a href=#ncbitax2lin class=md-nav__link> <span class=md-ellipsis> ncbitax2lin </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#diamond-index class=md-nav__link> <span class=md-ellipsis> DIAMOND index </span> </a> </li> <li class=md-nav__item> <a href=#creating-list-of-all-protein-accession-ids-in-nr class=md-nav__link> <span class=md-ellipsis> Creating list of all protein accession ids in NR </span> </a> <nav class=md-nav aria-label="Creating list of all protein accession ids in NR"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#download-entire-nr class=md-nav__link> <span class=md-ellipsis> Download entire nr </span> </a> </li> <li class=md-nav__item> <a href=#extract-all-accession-ids-from-the-nrgz-file class=md-nav__link> <span class=md-ellipsis> Extract all accession ids from the nr.gz file </span> </a> </li> <li class=md-nav__item> <a href=#split class=md-nav__link> <span class=md-ellipsis> Split </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#get-the-accession_id-to-taxid-lookup class=md-nav__link> <span class=md-ellipsis> Get the accession_id-to-taxid lookup </span> </a> <nav class=md-nav aria-label="Get the accession_id-to-taxid lookup"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#download-it-and-split-it class=md-nav__link> <span class=md-ellipsis> Download it, and split it </span> </a> </li> <li class=md-nav__item> <a href=#add-taxid-information-from-protaccession2taxidfullgz-to-the-nr-accid-splits class=md-nav__link> <span class=md-ellipsis> add taxid information from prot.accession2taxid.FULL.gz to the nr accid splits </span> </a> </li> <li class=md-nav__item> <a href=#update-unknowns class=md-nav__link> <span class=md-ellipsis> update "Unknowns" </span> </a> </li> <li class=md-nav__item> <a href=#create-lookup-table-pickle class=md-nav__link> <span class=md-ellipsis> create lookup table pickle </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#create-nr-titles-lookup-pickle class=md-nav__link> <span class=md-ellipsis> Create nr titles lookup pickle </span> </a> <nav class=md-nav aria-label="Create nr titles lookup pickle"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#extract-sequence-id-and-title class=md-nav__link> <span class=md-ellipsis> extract sequence id and title </span> </a> </li> <li class=md-nav__item> <a href=#build-pickle class=md-nav__link> <span class=md-ellipsis> build pickle </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#diamond-annotation-workflow class=md-nav__link> <span class=md-ellipsis> DIAMOND annotation workflow </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1>Resources</h1> <p>This pipeline requires a few resources to be created ahead of time to be supplied to the pipeline at runtime.</p> <h3 id=genomes>Genomes<a class=headerlink href=#genomes title="Permanent link">&para;</a></h3> <p>mm10 (mouse) and hg38 (human) genomes were downloaded from UCSC directly. Only canonical chromosomes were retained, that is, chr1 through chr19 plus chrX, chrY and chrM for mouse; and chr1 through chr22 plus chrX, chrY and chrM for human.</p> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span>wget<span class=w> </span>https://hgdownload.soe.ucsc.edu/goldenPath/mm10/bigZips/mm10.fa.gz
⠠⠵<span class=w> </span>wget<span class=w> </span>https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
⠠⠵<span class=w> </span>gzip<span class=w> </span>-d<span class=w> </span>hg38.fa.gz
⠠⠵<span class=w> </span>mv<span class=w> </span>hg38.fa<span class=w> </span>hg38.full.fa
⠠⠵<span class=w> </span>module<span class=w> </span>load<span class=w> </span>seqtk<span class=w> </span><span class=o>&amp;&amp;</span><span class=w> </span>seqtk<span class=w> </span>subseq<span class=w> </span>hg38.full.fa<span class=w> </span>chroms.lst<span class=w> </span>&gt;<span class=w> </span>hg38.fa
⠠⠵<span class=w> </span>cat<span class=w> </span>chroms.lst
chr1
chr2
chr3
chr4
chr5
chr6
chr7
chrX
chr8
chr9
chr11
chr10
chr12
chr13
chr14
chr15
chr16
chr17
chr18
chr20
chr19
chrY
chr22
chr21
chrM
</code></pre></div> <h3 id=star-index>STAR index<a class=headerlink href=#star-index title="Permanent link">&para;</a></h3> <p>STAR indices were built using version 2.7.9a and located at <code>/data/EVset_RNAseq/Pipelines/ERVPipeline/resources</code>.</p> <h3 id=geve-annotations>GEVE annotations<a class=headerlink href=#geve-annotations title="Permanent link">&para;</a></h3> <ul> <li>GTF file is copied from <code>/data/EVset_RNAseq/chidesters2/ERV_code/Hsap38.geve.m_v2_SRK.gtf</code> and hardlinked as <code>hg38.geve.gtf</code></li> <li>GTF file is copied from <code>/data/EVset_RNAseq/chidesters2/ERV_code/Mmus38.geve.m_v2_SRK.gtf</code> and hardlinked as <code>mm10.geve.gtf</code></li> <li><code>mm10.geve.annotation_table.tsv</code></li> </ul> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span>wget<span class=w> </span>//geve.med.u-tokai.ac.jp/download_data/table/Mmus38.txt.bz2
⠠⠵<span class=w> </span>bzip2<span class=w> </span>-d<span class=w> </span>Mmus38.txt.bz2
⠠⠵<span class=w> </span>ln<span class=w> </span>Mmus38.txt<span class=w> </span>mm10.geve.annotation_table.tsv
</code></pre></div> <ul> <li><code>hg38.geve.annotation_table.tsv</code></li> </ul> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span>wget<span class=w> </span>http://geve.med.u-tokai.ac.jp/download_data/table/Hsap38.txt.bz2
⠠⠵<span class=w> </span>bzip2<span class=w> </span>-d<span class=w> </span>Hsap38.txt.bz2
⠠⠵<span class=w> </span>ln<span class=w> </span>Hsap38.txt<span class=w> </span>hg38.geve.annotation_table.tsv
</code></pre></div> <h3 id=creating-taxid-to-lineage-lookup>Creating taxid-to-lineage lookup<a class=headerlink href=#creating-taxid-to-lineage-lookup title="Permanent link">&para;</a></h3> <h4 id=download-dmp-files>Download .dmp files<a class=headerlink href=#download-dmp-files title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span>wget<span class=w> </span>ftp://ftp.ncbi.nih.gov/pub/taxonomy/new_taxdump/new_taxdump.tar.gz<span class=w>  </span>.
⠠⠵<span class=w> </span>tar<span class=w> </span>xzvf<span class=w> </span>new_taxdump.tar.gz
</code></pre></div> <p>The following <code>.dmp</code> files are extracted:</p> <ul> <li>citations.dmp</li> <li>delnodes.dmp</li> <li>division.dmp</li> <li>excludedfromtype.dmp</li> <li>fullnamelineage.dmp</li> <li>gencode.dmp</li> <li>host.dmp</li> <li>images.dmp</li> <li>merged.dmp</li> <li>names.dmp</li> <li>nodes.dmp</li> <li>rankedlineage.dmp</li> <li>taxidlineage.dmp</li> <li>typematerial.dmp</li> <li>typeoftype.dmp</li> </ul> <p>We only need <code>names.dmp</code> and <code>nodes.dmp</code></p> <h4 id=ncbitax2lin>ncbitax2lin<a class=headerlink href=#ncbitax2lin title="Permanent link">&para;</a></h4> <p><a href=https://github.com/zyxue/ncbitax2lin>ncbitax2lin</a> converts the <code>names.dmp</code> and <code>nodes.dmp</code> to a lineage csv file with columns:</p> <ul> <li>tax_id</li> <li>superkingdom</li> <li>phylum</li> <li>class</li> <li>order</li> <li>family</li> <li>genus</li> <li>species</li> <li>biotype</li> <li>clade</li> <li>clade1</li> <li>clade10</li> <li>clade11</li> <li>clade12</li> <li>clade13</li> <li>clade14</li> <li>clade15</li> <li>clade16</li> <li>clade17</li> <li>clade18</li> <li>clade19</li> <li>clade2</li> <li>clade3</li> <li>clade4</li> <li>clade5</li> <li>clade6</li> <li>clade7</li> <li>clade8</li> <li>clade9</li> <li>cohort</li> <li>forma</li> <li>forma specialis</li> <li>forma specialis1</li> <li>genotype</li> <li>infraclass</li> <li>infraorder</li> <li>isolate</li> <li>kingdom</li> <li>morph</li> <li>no rank</li> <li>no rank1</li> <li>no rank2</li> <li>no rank3</li> <li>no rank4</li> <li>no rank5</li> <li>parvorder</li> <li>pathogroup</li> <li>section</li> <li>series</li> <li>serogroup</li> <li>serotype</li> <li>species group</li> <li>species subgroup</li> <li>strain</li> <li>subclass</li> <li>subcohort</li> <li>subfamily</li> <li>subgenus</li> <li>subkingdom</li> <li>suborder</li> <li>subphylum</li> <li>subsection</li> <li>subspecies</li> <li>subtribe</li> <li>superclass</li> <li>superfamily</li> <li>superorder</li> <li>superphylum</li> <li>tribe</li> <li>varietas</li> </ul> <p>ncbitax2lin is available in the "py311" conda env and requires very large amount of memory. H ence, submitted like:</p> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span>sbatch<span class=w> </span>slurm_job_high_memory<span class=w> </span>ncbitaxid2lineage.sh
</code></pre></div> <p>where</p> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span>more<span class=w> </span>slurm_job_high_memory
<span class=c1>#!/bin/csh -v</span>
<span class=c1>#SBATCH --job-name=&quot;taxid&quot;</span>
<span class=c1>#SBATCH --mail-type=BEGIN,END</span>
<span class=c1>#SBATCH --mem=1500g</span>
<span class=c1>#SBATCH --partition=&quot;largemem&quot;</span>
<span class=c1>#SBATCH --time=24:00:00</span>


<span class=nb>cd</span><span class=w> </span><span class=nv>$SLURM_SUBMIT_DIR</span>
sh<span class=w> </span><span class=nv>$1</span>
</code></pre></div> <p>and</p> <div class=highlight><pre><span></span><code><span class=ch>#!/bin/bash</span>
.<span class=w> </span><span class=s2>&quot;/data/CCBR_Pipeliner/db/PipeDB/Conda/etc/profile.d/conda.sh&quot;</span>
conda<span class=w> </span>activate<span class=w> </span>py311
ncbitax2lin<span class=w> </span>--nodes-file<span class=w> </span>nodes.dmp<span class=w> </span>--names-file<span class=w> </span>names.dmp<span class=w> </span>--output<span class=w> </span>ncbi_lineages.csv.gz
</code></pre></div> <h3 id=diamond-index>DIAMOND index<a class=headerlink href=#diamond-index title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span>module<span class=w> </span>load<span class=w> </span>diamond
⠠⠵<span class=w> </span>diamond<span class=w> </span>makedb<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--threads<span class=w> </span><span class=m>56</span><span class=w> </span><span class=se>\</span>
<span class=w>    </span>--in<span class=w> </span>nr.gz<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--db<span class=w> </span>nr<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--taxonmap<span class=w> </span>prot.accession2taxid.FULL.gz<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--taxonnodes<span class=w> </span>nodes.dmp<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--taxonnames<span class=w> </span>names.dmp
</code></pre></div> <ul> <li>creates <code>nr.dmnd</code></li> <li>it is about 300 GB in size</li> </ul> <h3 id=creating-list-of-all-protein-accession-ids-in-nr>Creating list of all protein accession ids in NR<a class=headerlink href=#creating-list-of-all-protein-accession-ids-in-nr title="Permanent link">&para;</a></h3> <h4 id=download-entire-nr>Download entire nr<a class=headerlink href=#download-entire-nr title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span>wget<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/blast/db/FASTA/nr.gz
</code></pre></div> <h4 id=extract-all-accession-ids-from-the-nrgz-file>Extract all accession ids from the <code>nr.gz</code> file<a class=headerlink href=#extract-all-accession-ids-from-the-nrgz-file title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span>zcat<span class=w> </span>nr.gz<span class=p>|</span>grep<span class=w> </span><span class=s2>&quot;^&gt;&quot;</span><span class=p>|</span>awk<span class=w> </span><span class=s1>&#39;{print $1}&#39;</span><span class=p>|</span>sed<span class=w> </span><span class=s2>&quot;s/&gt;//g&quot;</span><span class=w> </span>&gt;<span class=w> </span>nr.accession_ids
</code></pre></div> <p>This accounts to about 596 mil. accession ids. (~ 0.6 trillion)</p> <blockquote> <p>NOTE: some accession numbers have untrimmed prefixes or suffixes like: pir|E49255| prf||0803150A these should be E49255 0803150A</p> </blockquote> <h4 id=split>Split<a class=headerlink href=#split title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span>split<span class=w> </span>-d<span class=w> </span>-a<span class=w> </span><span class=m>5</span><span class=w> </span>-l<span class=w> </span><span class=m>1000000</span><span class=w> </span>nr.accession_ids<span class=w> </span>nr.accession_ids.
</code></pre></div> <p>This generates ~600 files.</p> <h3 id=get-the-accession_id-to-taxid-lookup>Get the accession_id-to-taxid lookup<a class=headerlink href=#get-the-accession_id-to-taxid-lookup title="Permanent link">&para;</a></h3> <h4 id=download-it-and-split-it>Download it, and split it<a class=headerlink href=#download-it-and-split-it title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span>wget<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.FULL.gz
⠠⠵<span class=w> </span>pigz<span class=w> </span>-d<span class=w> </span>-p<span class=w> </span><span class=m>4</span><span class=w> </span>prot.accession2taxid.FULL.gz
⠠⠵<span class=w> </span>split<span class=w> </span>-d<span class=w> </span>-a<span class=w> </span><span class=m>5</span><span class=w> </span>-l<span class=w> </span><span class=m>1000000</span><span class=w> </span>prot.accession2taxid.FULL<span class=w> </span>prot.accession2taxid.
</code></pre></div> <p>This creates 6356 files with 6355105289 total lines, i.e. ~ 6.4 trillion accession_id-to-taxid combinations.</p> <blockquote> <p>STATS:</p> <p>Total number of accession numbers in <code>nr.gz</code>: ~ 0.6 trillion</p> <p>Total number of protein accid to taxid rows in <code>prot.accession2taxid.FULL</code>: ~ 6.4 trillion</p> </blockquote> <h4 id=add-taxid-information-from-protaccession2taxidfullgz-to-the-nr-accid-splits>add taxid information from <code>prot.accession2taxid.FULL.gz</code> to the nr accid splits<a class=headerlink href=#add-taxid-information-from-protaccession2taxidfullgz-to-the-nr-accid-splits title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span><span class=k>for</span><span class=w> </span>i<span class=w> </span><span class=k>in</span><span class=w> </span><span class=o>{</span><span class=m>00000</span>..00595<span class=o>}</span><span class=p>;</span><span class=k>do</span><span class=w> </span><span class=nb>echo</span><span class=w> </span>bash<span class=w> </span>./<span class=w>   </span>add_taxid.sh<span class=w> </span>nr.accession_ids.splits/nr.accession_ids.
<span class=si>${</span><span class=nv>i</span><span class=si>}</span><span class=w> </span>nr.accession_ids.splits/nr.accession_ids.<span class=si>${</span><span class=nv>i</span><span class=si>}</span>.w_taxid<span class=p>;</span><span class=k>done</span><span class=w> </span>&gt;<span class=w> </span>do_it

⠠⠵<span class=w> </span>swarm<span class=w> </span>-f<span class=w> </span>do_it<span class=w> </span>-t<span class=w> </span><span class=m>2</span><span class=w> </span>-g<span class=w> </span><span class=m>120</span><span class=w> </span>--time<span class=o>=</span><span class=m>120</span>:00:00
</code></pre></div> <ul> <li>The <code>add_taxid.sh</code> is run for each nr.accession_ids.split. It iterates through each of the 6356 prot.accession2taxid splits and extracts the accid to taxid information. If taxid is not found the "Unknown" is reported.</li> <li>the <code>nr.accession_ids.?????.w_taxid</code> files contain a lot of "Unknown". (~ 13.5k per split) . Some "Unknown" can be re-queried with ncbi using <strong>edirect</strong> utils and taxid information can be distilled</li> </ul> <h4 id=update-unknowns>update "Unknowns"<a class=headerlink href=#update-unknowns title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span><span class=k>for</span><span class=w> </span>f<span class=w> </span><span class=k>in</span><span class=w> </span><span class=o>{</span><span class=m>00000</span>..00595<span class=o>}</span><span class=p>;</span><span class=k>do</span><span class=w> </span><span class=nb>echo</span><span class=w> </span><span class=s2>&quot;bash ./post_process.sh </span><span class=nv>$f</span><span class=s2> &gt; </span><span class=si>${</span><span class=nv>f</span><span class=si>}</span><span class=s2>.log 2&gt;&amp;1&quot;</span><span class=p>;</span><span class=k>done</span><span class=w> </span>&gt;<span class=w> </span>do_update

⠠⠵<span class=w> </span>swarm<span class=w> </span>-f<span class=w> </span>do_update<span class=w> </span>-t<span class=w> </span><span class=m>2</span><span class=w> </span>-m<span class=w> </span>120g<span class=w> </span>--time<span class=o>=</span><span class=m>1</span>:59:00<span class=w> </span>-m<span class=w> </span>edirect<span class=w> </span>--exclusive<span class=w> </span>--maxrunning<span class=w> </span><span class=m>1</span><span class=w> </span>--partition<span class=w> </span>quick
</code></pre></div> <ul> <li> <p><code>post_process.sh</code></p> </li> <li> <p>extracts "Unknown"s from the <code>*.w_taxid</code> files</p> </li> <li>splits the "Unknown"s into chunks of 100 each</li> <li>submit a query to NCBI using esummary with 100 accids at a time</li> <li>gather all new taxid information</li> <li> <p>save <code>*.w_taxid.updated</code> file for each nr.accession_ids split</p> </li> <li> <p>"Unknown"s now average to about 178 per nr.accession_ids split. There are a total of 106138 "Unknowns" in the entire ~0.6 trillion nr.accession_ids. Most of these are due to deprecated accession ids or proteins no longer being part of any NCBI genome. These can be safely ignored.</p> </li> </ul> <h4 id=create-lookup-table-pickle>create lookup table pickle<a class=headerlink href=#create-lookup-table-pickle title="Permanent link">&para;</a></h4> <p>All the 596 updated files can be read into a single dict and saved as pickle for future 1-to-1 access. Memory needed to load the entire lookup table is around 130GB. The pkl file size is around 15GB.</p> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span>python<span class=w> </span>topkl.py<span class=w> </span>accid2taxid.pkl
</code></pre></div> <ul> <li><code>accid2taxid.pkl</code> is 14GB in size</li> <li>loading the entire pkl in memory takes:</li> <li>~ 15 min</li> <li>~ 120 GB of RAM</li> </ul> <h3 id=create-nr-titles-lookup-pickle>Create nr titles lookup pickle<a class=headerlink href=#create-nr-titles-lookup-pickle title="Permanent link">&para;</a></h3> <h4 id=extract-sequence-id-and-title>extract sequence id and title<a class=headerlink href=#extract-sequence-id-and-title title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span>zcat<span class=w> </span>nr.gz<span class=p>|</span>grep<span class=w> </span><span class=s2>&quot;^&gt;&quot;</span><span class=p>|</span>sed<span class=w> </span><span class=s2>&quot;s/&gt;//g&quot;</span><span class=w> </span>&gt;<span class=w> </span>nr.ids
</code></pre></div> <h4 id=build-pickle>build pickle<a class=headerlink href=#build-pickle title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code>⠠⠵<span class=w> </span>python<span class=w> </span>nr_topkl.py<span class=w> </span>nr_titles.pkl
</code></pre></div> <ul> <li><code>nr_titles.pkl</code> is ~ 72GB is size</li> <li>requires about 200 GB of RAM to load</li> </ul> <h3 id=diamond-annotation-workflow>DIAMOND annotation workflow<a class=headerlink href=#diamond-annotation-workflow title="Permanent link">&para;</a></h3> <p><img alt=workflow src=../assets/images/diamond_annotation_workflow.png></p> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg> </span> 2024-01-30 </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2022 CCBR </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=http://bioinformatics.cancer.gov target=_blank rel=noopener title=bioinformatics.cancer.gov class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M144 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160zm368 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160zM0 298.7C0 239.8 47.8 192 106.7 192h42.7c15.9 0 31 3.5 44.6 9.7-1.3 7.2-1.9 14.7-1.9 22.3 0 38.2 16.8 72.5 43.3 96H21.3C9.6 320 0 310.4 0 298.7zM405.3 320h-.7c26.6-23.5 43.3-57.8 43.3-96 0-7.6-.7-15-1.9-22.3 13.6-6.3 28.7-9.7 44.6-9.7h42.7c58.9 0 106.7 47.8 106.7 106.7 0 11.8-9.6 21.3-21.3 21.3H405.3zM224 224a96 96 0 1 1 192 0 96 96 0 1 1-192 0zm-96 261.3c0-73.6 59.7-133.3 133.3-133.3h117.4c73.6 0 133.3 59.7 133.3 133.3 0 14.7-11.9 26.7-26.7 26.7H154.7c-14.7 0-26.7-11.9-26.7-26.7z"/></svg> </a> <a href=https://github.com/CCBR target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://hub.docker.com/orgs/nciccbr/repositories target=_blank rel=noopener title=hub.docker.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M349.9 236.3h-66.1v-59.4h66.1v59.4zm0-204.3h-66.1v60.7h66.1V32zm78.2 144.8H362v59.4h66.1v-59.4zm-156.3-72.1h-66.1v60.1h66.1v-60.1zm78.1 0h-66.1v60.1h66.1v-60.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1l-13.3-8.9zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm-78.1-72.1h-66.1v60.1h66.1v-60.1z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "..", "features": ["navigation.tabs", "navigation.top", "navigation.indexes", "toc.integrate"], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../assets/javascripts/bundle.e1c3ead8.min.js></script> </body> </html>